Testing

Ki11erRabbit opened the pull request:
I have written a way to generate a bunch of tests from a Rust source file. It is designed to be used in conjunction with something else to actually run the tests and check the outputs. There are a few parts that are a little rough so feel free to suggest changes to the code. It could use some optimizations but for a program that has a short lifespan I think it should be acceptable.

(Potential) Issues
* Functions are too vague
* I think I got SubrFn right

CeleritasCelery commented:
This looks great! I took some time to play around with it. A couple of thoughts

1. It panics if it encounters an unknown type. Instead we should print an error message and than move on to the next function.
2. We should find some way to remove duplicates results. Most of the generated types tend to be the same.
3. Have we tried [deriving arbitrary](https://docs.rs/arbitrary/latest/arbitrary/trait.Arbitrary.html#deriving-arbitrary) instead of implementing it ourselves?
4. related to that, we seem to have duplicate logic between `impl<'a> Arbitrary<'a> for ArbitraryObjectType` and `impl From<ObjectType> for ArbitraryObjectType`. Could we just have one implementation of logic?
5. It would be valuable to pass in the function you want to test as an argument. That way if you implement a single function you can test just that one.
6. The previous code had some test runners to run the functions against GNU Emacs. Is that something you want to include in this or are you thinking that should be part of a different binary?

I am really happy with how this is looking. Let me know when you want me to merge it, I am okay with merging code that is not "complete" so long as you feel ready to commit it.

Ki11erRabbit commented:
I'll clean up the code a little and implement your thoughts.
The reason why Arbitrary is not being derived is because it sometimes doesn't create interesting tests or it causes tests to take forever to print. An example for each would be strings and lists (it can create invalid symbols or empty lists) and for functions it tends to generate arity 50+ which is unreasonable.

Ki11erRabbit commented:
I have added a way to filter functions. Not too sure on how to remove duplicates. I could make a wrapper for `f64` that can implement `Hash` so that we can just use a hashset.
I also moved the logic for the `Arbitrary` and `From` into helper functions.

 I think that there should be a separate runner for testing since I think that it would make maintenance easier in the long run.
